src/allmydata/storage_client.py

```diff
@@ -1227,35 +1217,29 @@ def request(reactor, nurl: DecodedURL):
             finally:
                 self._connecting_deferred = None
 
-            if nurl is None:
-                # We failed to find a server to connect to. Perhaps the next
-                # iteration of the loop will succeed.
-                return
-            else:
-                self._istorage_server = _HTTPStorageServer.from_http_client(
-                    StorageClient.from_nurl(nurl, reactor)
-                )
-
-        result = self._istorage_server.get_version()
-
-        def remove_connecting_deferred(result):
-            self._connecting_deferred = None
-            return result
-
-        # Set a short timeout since we're relying on this for server liveness.
-        self._connecting_deferred = result.addTimeout(5, self._reactor).addBoth(
-                remove_connecting_deferred).addCallbacks(
-            self._got_version,
-            self._failed_to_connect
-        )
+            # If we've gotten this far, we've found a working NURL.
+            self._istorage_server = _HTTPStorageServer.from_http_client(
+                StorageClient.from_nurl(nurl, reactor)
+            )
+            return self._istorage_server
 
-        # We don't want to do another iteration of the loop until this
-        # iteration has finished, so wait here:
         try:
-            if self._connecting_deferred is not None:
-                await self._connecting_deferred
-        except Exception as e:
-            log.msg(f"Failed to connect to a HTTP storage server: {e}", level=log.CURIOUS)
+            try:
```

Comment:
The two layers of `try` could be collapsed into one, if you felt like it, since:

```
try:
  ...
except:
  ...
finally:
  x
```

executes `x` whether or not the `except` block is executed

Reply:
Ah, I'd forgotten about that.